File: SampleApplications/InteractionBasedTesting/Rhino.Mocks.Demo.CS/Order.cs
using System;
using System.Data.SqlClient;
using System.Runtime.CompilerServices;
[assembly: InternalsVisibleTo(Rhino.Mocks.RhinoMocks.StrongName)]


namespace Rhino.Mocks.Demo
{
	public class Order
	{
		private decimal total;

		public decimal Total
		{
			get { return total; }
			set { total = value; }
		}

		public decimal GetTotal()
		{
			decimal salesHistory;
			salesHistory = GetSalesHistory();
			if (salesHistory >= 1000)
			{
				if (total >= 100)
				{
					return total - (total * 0.10m);
				}
				else if (total >= 60)
				{
					return total - (total * 0.05m);
				}
			}
			return total;
		}

		protected internal virtual decimal GetSalesHistory()
		{
			decimal salesHistory;
			using (SqlConnection sqlConnection = new SqlConnection("dummy"))
			{
				sqlConnection.Open();
				salesHistory = DataAccess.GetSalesHistory(sqlConnection, this);
			}
			return salesHistory;
		}
	}

	public static class DataAccess
	{
		static Random random = new Random();
		public static decimal GetSalesHistory(SqlConnection connection,
			Order order)
		{
			return random.Next(10, 10000);
		}
	}
}

File: SampleApplications/InteractionBasedTesting/Rhino.Mocks.Demo.VB/Order.vb
Imports System
Imports System.Data.SqlClient
Imports System.Runtime.CompilerServices

Public Class Order
	Private m_total As Decimal

	Public Property Total() As Decimal
		Get
			Return m_total
		End Get
		Set(ByVal value As Decimal)
			m_total = value
		End Set
	End Property

	Public Function GetTotal() As Decimal
		Dim salesHistory As Decimal
		salesHistory = GetSalesHistory()
		If salesHistory >= 1000 Then
			If m_total >= 100 Then
				Return m_total - (m_total * 0.1D)
			ElseIf m_total >= 60 Then
				Return m_total - (m_total * 0.05D)
			End If
		End If
		Return m_total
	End Function

	Protected Friend Overridable Function GetSalesHistory() As Decimal
		Dim salesHistory As Decimal
		Using sqlConnection As New SqlConnection("dummy")
			sqlConnection.Open()
			salesHistory = DataAccess.GetSalesHistory(sqlConnection, Me)
		End Using
		Return salesHistory
	End Function
End Class

Public NotInheritable Class DataAccess
	Private Sub New()
	End Sub
	Shared random As New Random()
	Public Shared Function GetSalesHistory(ByVal connection As SqlConnection, ByVal order As Order) As Decimal
		Return random.[Next](10, 10000)
	End Function
End Class

